core/src/main/scala/chisel3/internal/firrtl/IR.scala

```diff
@@ -861,7 +861,14 @@ case class DefBlackBox(
   params: Map[String, Param])
     extends Component
 
-case class Circuit(name: String, components: Seq[Component], annotations: Seq[ChiselAnnotation], renames: RenameMap) {
-  def firrtlAnnotations: Iterable[Annotation] = annotations.flatMap(_.toFirrtl.update(renames))
-
+case class Circuit(
+  name:           String,
+  components:     Seq[Component],
+  annotations:    Seq[ChiselAnnotation],
+  renames:        RenameMap,
+  newAnnotations: Seq[ChiselToFirrtlAnnotations] = Seq.empty) {
```

Comment:
Adding fields to case classes is both a binary incompatible and source incompatible change. This would not be an issue if we had properly made all of these classes private.... but we didn't and [some users do notice](https://github.com/chipsalliance/chisel3/issues/2307).

To safely backport this to 3.5.x (which I presume is your plan) we either need to [use dataclass](https://github.com/chipsalliance/firrtl/pull/2071/files) to give case class-like behavior in a more friendly way for making modifications like this, or we need to add these new arguments to a 2nd parameter list and implement equals and hashcode ourselves.

Reply:
@jackkoenig  let me explore that - I am assuming since the `annotations` variable is publicly visible - the change suggested by @seldridge  to make `annotations`  `Seq[ChiselToFirrtlAnnotations]` instead of `Seq[ChiselAnnotation]` also is disruptive?

Reply:
> @jackkoenig let me explore that - I am assuming since the `annotations` val is publicly visible - the change suggested by @seldridge to make `annotations` `Seq[ChiselToFirrtlAnnotations]` instead of `Seq[ChiselAnnotation]` also is disruptive?

Reply:
@jackkoenig  please take a look - have used dataclass - however not sure if I have handled copy properly (also do we need an unapply method? )

Reply:
Copy should exclude any new methods (any after `@since`), also please add `"Chisel 3.5.4"` to the since just to make it obvious. You also do need to define unapply. To check if your changes are binary compatible, you can try making them against 3.5.x and then running `sbt +mimaReportBinaryIncompatibilities`

See https://github.com/chipsalliance/firrtl/blob/c7403e80f965aa3c6f65b176dbb56dbb905cbb31/src/main/scala/firrtl/ir/IR.scala#L769 for example of the copy not having any new fields and of implementing unapply yourself (again not including the new fields).

Reply:
@jackkoenig did what you suggested - and I am seeing similar errors as mentioned in the PR you linked after my update - can these be ignored? 
```
error] chisel3-core: Failed binary compatibility check against edu.berkeley.cs:chisel3-core_2.12:3.5.3! Found 4 potential problems (filtered 8)
[error]  * class chisel3.internal.firrtl.Circuit is declared final in current version
[error]    filter with: ProblemFilters.exclude[FinalClassProblem]("chisel3.internal.firrtl.Circuit")
[error]  * static method tupled()scala.Function1 in class chisel3.internal.firrtl.Circuit does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("chisel3.internal.firrtl.Circuit.tupled")
[error]  * static method curried()scala.Function1 in class chisel3.internal.firrtl.Circuit does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("chisel3.internal.firrtl.Circuit.curried")
[error]  * the type hierarchy of object chisel3.internal.firrtl.Circuit is different in current version. Missing types {scala.runtime.AbstractFunction4}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("chisel3.internal.firrtl.Circuit$")

```

Reply:
@jackkoenig have incorporated your changes to resolve the compatibility issues - the check is still running locally - and so far so good - will update once it ends (hopefully passes).

Reply:
It passed! 

```
sbt +mimaReportBinaryIssues
....
[success] Total time: 22 s, completed Jul 13, 2022 11:25:27 AM
[info] Setting Scala version to 2.13.7 on 6 projects.
[info] Reapplying settings...
[info] set current project to chisel3 (in build file:/scratch/girishp/vscode/clone_23jun/federation/rocket-chip/chisel3/)
[success] Total time: 0 s, completed Jul 13, 2022 11:25:28 AM
[info] Setting Scala version to 2.13.8 on 6 projects.
[info] Reapplying settings...
[info] set current project to chisel3 (in build file:/scratch/girishp/vscode/clone_23jun/federation/rocket-chip/chisel3/)
[success] Total time: 0 s, completed Jul 13, 2022 11:25:29 AM
[info] Reapplying settings...
[info] set current project to chisel3 (in build file:/scratch/girishp/vscode/clone_23jun/federation/rocket-chip/chisel3/)
```

Reply:
This is on 3.5.x ....